File: app/src/main/java/com/vwoom/timelapsegallery/widget/WidgetGridRemoteViewsFactory.java
package com.vwoom.timelapsegallery.widget;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.os.Bundle;
import android.util.Log;
import android.widget.RemoteViews;
import android.widget.RemoteViewsService;

import com.vwoom.timelapsegallery.R;
import com.vwoom.timelapsegallery.database.entry.CoverPhotoEntry;
import com.vwoom.timelapsegallery.database.entry.PhotoEntry;
import com.vwoom.timelapsegallery.database.entry.ProjectEntry;
import com.vwoom.timelapsegallery.database.TimeLapseDatabase;
import com.vwoom.timelapsegallery.database.entry.ProjectScheduleEntry;
import com.vwoom.timelapsegallery.utils.FileUtils;
import com.vwoom.timelapsegallery.utils.Keys;
import com.vwoom.timelapsegallery.utils.PhotoUtils;
import com.vwoom.timelapsegallery.utils.TimeUtils;

import java.io.IOException;
import java.util.List;

public class WidgetGridRemoteViewsFactory implements RemoteViewsService.RemoteViewsFactory {

    private Context mContext;
    private List<ProjectEntry> mProjects;
    private TimeLapseDatabase mTimeLapseDatabase;
    private static final String TAG = WidgetGridRemoteViewsFactory.class.getSimpleName();

    public WidgetGridRemoteViewsFactory(Context applicationContext, Intent intent){
        mContext = applicationContext;
        mProjects = null;
        mTimeLapseDatabase = TimeLapseDatabase.getInstance(mContext);
    }

    @Override
    public void onCreate() {

    }

    @Override
    public void onDataSetChanged() {
        /* Load the projects for the day */
        mProjects = mTimeLapseDatabase.projectDao().loadAllScheduledProjects();

    }

    @Override
    public void onDestroy() {

    }

    @Override
    public int getCount() {
        if (mProjects == null) return 0;
        else return mProjects.size();
    }

    @Override
    public RemoteViews getViewAt(int i) {
        // Get the current project
        ProjectEntry currentProject = mProjects.get(i);
        ProjectScheduleEntry schedule = mTimeLapseDatabase.projectScheduleDao().loadScheduleByProjectId(currentProject.getId());
        CoverPhotoEntry coverPhotoEntry = mTimeLapseDatabase.coverPhotoDao().getCoverPhoto_nonLiveData(currentProject.getId());
        PhotoEntry coverPhoto = mTimeLapseDatabase.photoDao().loadPhoto(currentProject.getId(), coverPhotoEntry.getPhoto_id());

        long nextSubmissionTime = TimeUtils.getNextScheduledSubmission(schedule.getSchedule_time(), schedule.getInterval_days());

        // Get strings
        String nextSubmissionTimeString = TimeUtils.getTimeFromTimestamp(nextSubmissionTime);

        // Create the remote views
        RemoteViews views = new RemoteViews(mContext.getPackageName(), R.layout.widget_list_item);

        String coverPhotoPath = FileUtils.getPhotoUrl(mContext, currentProject, coverPhoto);
        // Decode the bitmap from path
        Bitmap bitmap = PhotoUtils.decodeSampledBitmapFromPath(
                coverPhotoPath,
                100,
                100);

        // Rotate the bitmap
        try {
            Integer bitmapOrientation = PhotoUtils.getOrientationFromImagePath(coverPhotoPath);
            bitmap = PhotoUtils.rotateBitmap(bitmap, bitmapOrientation);
        } catch (IOException e){
            if (e.getMessage()!=null)
                Log.e(TAG, e.getMessage());
        }

        // Set the view strings
        views.setTextViewText(R.id.widget_list_item_name_text_view, currentProject.getProject_name());
        views.setTextViewText(R.id.widget_list_item_time_text_view, nextSubmissionTimeString);
        views.setImageViewBitmap(R.id.widget_list_item_image_view, bitmap);

        // Send the project in an intent
        Bundle extras = new Bundle();
        extras.putParcelable(Keys.PROJECT_ENTRY, currentProject);
        Intent fillInIntent = new Intent();
        fillInIntent.putExtras(extras);
        views.setOnClickFillInIntent(R.id.widget_grid_item_layout, fillInIntent);

        return views;
    }

    @Override
    public RemoteViews getLoadingView() {
        return null;
    }

    @Override
    public int getViewTypeCount() {
        return 1;
    }

    @Override
    public long getItemId(int i) {
        return i;
    }

    @Override
    public boolean hasStableIds() {
        return true;
    }

}

File: app/src/main/java/com/vwoom/timelapsegallery/widget/WidgetGridRemoteViewsFactory.kt
package com.vwoom.timelapsegallery.widget

import android.content.Context
import android.content.Intent
import android.graphics.Bitmap
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.widget.RemoteViews
import android.widget.RemoteViewsService.RemoteViewsFactory
import com.vwoom.timelapsegallery.R
import com.vwoom.timelapsegallery.data.entry.CoverPhotoEntry
import com.vwoom.timelapsegallery.data.entry.PhotoEntry
import com.vwoom.timelapsegallery.data.entry.ProjectEntry
import com.vwoom.timelapsegallery.data.repository.CoverPhotoRepository
import com.vwoom.timelapsegallery.data.repository.PhotoRepository
import com.vwoom.timelapsegallery.data.repository.ProjectRepository
import com.vwoom.timelapsegallery.data.repository.ProjectScheduleRepository
import com.vwoom.timelapsegallery.utils.FileUtils.getPhotoUrl
import com.vwoom.timelapsegallery.utils.PhotoUtils.decodeSampledBitmapFromPath
import com.vwoom.timelapsegallery.utils.PhotoUtils.getOrientationFromImagePath
import com.vwoom.timelapsegallery.utils.PhotoUtils.rotateBitmap
import kotlinx.coroutines.runBlocking
import java.io.File
import java.io.IOException

class WidgetGridRemoteViewsFactory(
        private val mContext: Context,
        private val projectRepository: ProjectRepository,
        private val coverPhotoRepository: CoverPhotoRepository,
        private val photoRepository: PhotoRepository) : RemoteViewsFactory {

    private var mProjects: List<ProjectEntry>
    private var mExternalFilesDir: File? = null

    override fun onCreate() {}

    override fun onDataSetChanged() {
        // Load the projects for the day
        mProjects = runBlocking { projectRepository.getScheduledProjects() }
    }

    override fun onDestroy() {}
    override fun getCount(): Int {
        return mProjects.size
    }

    override fun getViewAt(i: Int): RemoteViews { // Get the current project
        val currentProject: ProjectEntry = mProjects[i]
        val coverPhoto: CoverPhotoEntry = coverPhotoRepository.getCoverPhoto(currentProject.id)
        val photoEntry: PhotoEntry = photoRepository.getPhoto(currentProject.id, coverPhoto.photo_id)

        // Create the remote views
        val views = RemoteViews(mContext.packageName, R.layout.widget_grid_item)
        val coverPhotoPath = getPhotoUrl(mExternalFilesDir!!, currentProject, photoEntry.timestamp)
        // Decode the bitmap from path
        var bitmap: Bitmap? = decodeSampledBitmapFromPath(
                coverPhotoPath,
                200,
                200)
        // Rotate the bitmap
        try {
            val bitmapOrientation = getOrientationFromImagePath(coverPhotoPath)
            bitmap = rotateBitmap(bitmap!!, bitmapOrientation)
        } catch (e: IOException) {
            if (e.message != null) Log.e(TAG, e.message!!)
        }
        // Set the view strings
        views.setImageViewBitmap(R.id.widget_grid_item_image_view, bitmap)
        // Send the project in an intent
        val extras = Bundle()

        // TODO: determine what I want to do with this dang widget!
        //extras.putParcelable(Keys.PROJECT, currentProject);
        val fillInIntent = Intent()
        fillInIntent.putExtras(extras)
        views.setOnClickFillInIntent(R.id.widget_grid_item_image_view, fillInIntent)
        return views
    }

    override fun getLoadingView(): RemoteViews? {
        return null
    }

    override fun getViewTypeCount(): Int {
        return 1
    }

    override fun getItemId(i: Int): Long {
        return i.toLong()
    }

    override fun hasStableIds(): Boolean {
        return true
    }

    companion object {
        private val TAG = WidgetGridRemoteViewsFactory::class.java.simpleName
    }

    init {
        mExternalFilesDir = mContext.getExternalFilesDir(Environment.DIRECTORY_PICTURES)
        mProjects = arrayListOf()
    }
}